mypy/semanal.py

```diff
@@ -160,6 +160,45 @@
 Tag = int
 
 
+def substitute_unbounds(t: UnboundType, names_to_types: Mapping[str, Instance]) -> ProperType:
```

Comment:
Sorry for misleading suggestion, doing the copy during semantic analyzis is still bad. I was thinking about superclass being in a different module that is type-checked before the module with subclass, but it obviously doesn't work always.

Also this function is just one big hack. I was thinking about something along the lines of `mypy.checker.TypeChecker.bind_and_map_method()`.

For the latter to work well I see a possible strategy:

0. Add two flags (or maybe better one three state flag) to functions indicating whether this function can be "improved", i.e. it has some untyped arguments, while a supertype has those typed, and whether the "copying" has already been performed.
1. The first flag should be set during semantic analysis, this should be thought carefully to work well in import cycles.
2. During type checking, whenever we encounter a reference to a method `Sub.meth` for which we know it can be "improved", but wasn't yet, we trigger a deferral.
3. Finally, we perform the copying while analyzing the method itself (in a place where you did initially, but with a more sophisticated logic like currently)
4. Be sure that `aststrip.py` resets the flags, and `deps.py` generates dependencies so that subclass method callsites are all rechecked when the signature in supertype changes (this may be already the case, at least partially).

Reply:
@gantsevdenis I can't find your comment that I have got by e-mail, but here is the response. Self-types is something like this:
```python
T = TypeVar('T')
class Base:
    def copy(self: T) -> T: ...

class Sub(Base): ...
x = Sub().copy()  # inferred type of 'x' is 'Sub'
```